pkg/bootstrap/config.go

```diff
@@ -118,16 +118,18 @@ func (cfg Config) toTemplateParams() (map[string]any, error) {
 	opts = append(opts, getNodeMetadataOptions(cfg.Node)...)
 
 	// Check if nodeIP carries IPv4 or IPv6 and set up proxy accordingly
-	if network.AllIPv6(cfg.Metadata.InstanceIPs) {
-		opts = append(opts,
-			option.Localhost(option.LocalhostIPv6),
-			option.Wildcard(option.WildcardIPv6),
-			option.DNSLookupFamily(option.DNSLookupFamilyIPv6))
-	} else {
+	if network.AllIPv4(cfg.Metadata.InstanceIPs) {
+		// IPv4 only
 		opts = append(opts,
 			option.Localhost(option.LocalhostIPv4),
 			option.Wildcard(option.WildcardIPv4),
 			option.DNSLookupFamily(option.DNSLookupFamilyIPv4))
+	} else {
+		// IPv6 only and Dual Stack
+		opts = append(opts,
+			option.Localhost(option.LocalhostIPv6),
+			option.Wildcard(option.WildcardIPv6),
+			option.DNSLookupFamily(option.DNSLookupFamilyIPv6))
```

Comment:
Hi @howardjohn, it's okay for IPv6 only to add this option properties. But for dual stack, there are 2 situations:
1. [IPv4, IPv6]: it's working to deploy Istio by following the old code to set the DNSLookupFamilyIPv4
2. [IPv6, IPv4]: it's not working
However, we can uniform it by setting `DNSLookupFamilyIPv6` for IPv6 only and Dual Stack. Does it make sense?

Reply:
Why not auto?

Reply:
I think we use different logic in cluster_builder,does it make sense to differ?

Reply:
Yes, auto would be useful.  
But I have encountered an issue on different orders of IPs. The k8s cluster can be deployed in different ip order:

IPv4 first - by default
```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  ipFamily: dual
  podSubnet: "10.244.0.0/16, 2001:db8:42:0::/56"
  serviceSubnet: "10.96.0.0/12, 2001:db8:42:1::/112"
```

IPv6 first:
```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  ipFamily: dual
  podSubnet: "2001:db8:42:0::/56,10.244.0.0/16"
  serviceSubnet: "2001:db8:42:1::/112,10.96.0.0/12"
```
However, for `DNSLookupFamilyIP` policy setting, its order is fixed.

Reply:
> I think we use different logic in cluster_builder,does it make sense to differ?

This is helpful, besides the DNSLookupFamilyIP setting, there is a bug for cluster builder for UpstreamBindConfig

Reply:
The clusters being generate here may be for external addresses I think, by the way. so we cannot rely only on internal k8s config

Reply:
Its not clear to me why we use IPv6 only for Dual stack. What if I want to connect to github.com (no IPV6)?

Reply:
> The clusters being generate here may be for external addresses I think, by the way. so we cannot rely only on internal k8s config

Hi @howardjohn, the clusters generated like this may be for IPv6 applications access as well.  And we generate the `DNSLookupFamilyIP` configuration based on the IP family policies. I have re-submitted for it.

Reply:
> Its not clear to me why we use IPv6 only for Dual stack. What if I want to connect to github.com (no IPV6)?

Hi @howardjohn, according to the newest submitting, there are 3 branches for this:
1. IPv4 only
2. IPv6 only
3. Dual Stack
For 3, I kept the original logic for option's properties if dual stack is disable. And there also are 2 sub-branches for different IP family order. Does it make sense?
And for `What if I want to connect to github.com (no IPV6)?`
For upstream cluster building, there should be 2 bind config for it if source pods are deployed in Dual Stack cluster, so it should be okay because it has both IPv4 and IPv6 addresses. Hope my understand is correct and helpful to answer your question.

Reply:
sorry, I still don't get it. This configures the bootstrap clusters. Why do we want that to be different from the dynamic clusters? https://github.com/istio/istio/blob/373fc89518c986c9f48ed3cd891930da6fdc8628/pilot/pkg/networking/core/v1alpha3/cluster_builder.go#L370

They are used for the same purpose, by the same proxies. Shouldn't we be consistent?

Reply:
Hi @howardjohn, they are very similar, however there are some different, I can help to explain: 
1. The `DnsLookupFamily` setting in `cluster_builder.go` is for cluster DNS, and this PR is for basically for every pod's sidercar  `DnsLookupFamily`
2. They use different implementation method for the IP family checking, `SupportsIPv4` of proxy is used in `cluster_builder.go` and `AllIPv4`/`AllIPv6` of proxy are used in this PR. So they are semantically different.

And another thing is that there is `UpstreamBindConfig` for helping to smartly pick the right endpoint/pod even the default cluster DNS is IPv4. Does it make sense?

Reply:
> The DnsLookupFamily setting in cluster_builder.go is for cluster DNS, and this PR is for basically for every pod's sidercar DnsLookupFamily

It's the same field in envoy

Reply:
> It's the same field in envoy
Yeah, it's just for `STRICT_DNS` cluster, such as `zipkin`.

Reply:
> It's the same field in envoy
Yeah, it's just for `STRICT_DNS` cluster, such as `zipkin`. If so, it's okay to make both 2 places are consistent. Thanks John!

Reply:
> It's the same field in envoy

Yeah, it's just for `STRICT_DNS` cluster, such as `zipkin`. If so, it's okay to make both 2 places are consistent. Thanks John!

Reply:
Hi @howardjohn, re-submitted for cluster's DnsLookupFamily setting

Reply:
I see the change but cluster_builder logic still doesn't match this file. They are the exact same field in Envoy so I don't see why they should be different

Reply:
Hi @howardjohn, I have updated `DNSLookupFamilyIPv6` from `AUTO` to `V6_ONLY` and added `DNSLookupFamilyIPS` as `ALL` for dual stack case. Re-submitted! Thanks!